cpp/test/CMakeLists.txt

```diff
@@ -14,172 +14,308 @@
-)
-
-set_target_properties(test_raft
-        PROPERTIES BUILD_RPATH                         "\$ORIGIN"
-        # set target compile options
-        CXX_STANDARD                        17
-        CXX_STANDARD_REQUIRED               ON
-        CUDA_STANDARD                       17
-        CUDA_STANDARD_REQUIRED              ON
-        POSITION_INDEPENDENT_CODE           ON
-        INTERFACE_POSITION_INDEPENDENT_CODE ON
-        INSTALL_RPATH "\$ORIGIN/../../../lib"
-        )
-
-target_compile_options(test_raft
-        PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${RAFT_CXX_FLAGS}>"
-        "$<$<COMPILE_LANGUAGE:CUDA>:${RAFT_CUDA_FLAGS}>"
-        )
-
-target_include_directories(test_raft
-        PUBLIC  "$<BUILD_INTERFACE:${RAFT_SOURCE_DIR}/test>"
-        )
-
-
-target_link_libraries(test_raft
-        PRIVATE
-        raft::raft
-        raft::distance
-        raft::nn
-        GTest::gtest
-        GTest::gtest_main
-        Threads::Threads
-        $<TARGET_NAME_IF_EXISTS:OpenMP::OpenMP_CXX>
-        $<TARGET_NAME_IF_EXISTS:conda_env>)
-
-install(
-        TARGETS test_raft
-        COMPONENT testing
-        DESTINATION bin/libraft/gtests
-        EXCLUDE_FROM_ALL
-)
+
+###################################################################################################
+# - compiler function -----------------------------------------------------------------------------
+
+function(ConfigureTest CMAKE_TEST_NAME)
+    add_executable(${CMAKE_TEST_NAME} ${ARGN})
+
+    target_link_libraries(${CMAKE_TEST_NAME}
+            PRIVATE
+            raft::raft
+            GTest::gtest
+            GTest::gtest_main
+            Threads::Threads
+            $<TARGET_NAME_IF_EXISTS:OpenMP::OpenMP_CXX>
+            $<TARGET_NAME_IF_EXISTS:conda_env>
+            )
+
+
+    add_test(NAME ${CMAKE_TEST_NAME} COMMAND ${CMAKE_TEST_NAME})
+
+    set_target_properties(${CMAKE_TEST_NAME}
+            PROPERTIES BUILD_RPATH                         "\$ORIGIN"
+            # set target compile options
+            CXX_STANDARD                        17
+            CXX_STANDARD_REQUIRED               ON
+            CUDA_STANDARD                       17
+            CUDA_STANDARD_REQUIRED              ON
+            POSITION_INDEPENDENT_CODE           ON
+            INTERFACE_POSITION_INDEPENDENT_CODE ON
+            INSTALL_RPATH "\$ORIGIN/../../../lib"
+            )
+
+    target_compile_options(${CMAKE_TEST_NAME}
+            PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${RAFT_CXX_FLAGS}>"
+            "$<$<COMPILE_LANGUAGE:CUDA>:${RAFT_CUDA_FLAGS}>"
+            )
+
+    install(
+            TARGETS ${CMAKE_TEST_NAME}
+            COMPONENT testing
+            DESTINATION bin/gtests/libraft
+            EXCLUDE_FROM_ALL)
+endfunction()
+
+
+function(ConfigureDistanceTest CMAKE_TEST_NAME)
+    add_executable(${CMAKE_TEST_NAME} ${ARGN})
+
+    target_link_libraries(${CMAKE_TEST_NAME}
+            PRIVATE
+            raft::raft
+            raft::distance
```

Comment:
I think you could avoid the definitions of `ConfigureDistanceTest` and `ConfigureNearestNeighborsTest` by adding boolean parameters to `ConfigureTest`, which could save a lot of lines of code in this file.

Reply:
We do that in cuML for nccl, cumlprims amd others in https://github.com/rapidsai/cuml/blob/6f42ced4d6f84cc04ea27f8940449d57fa08ab51/cpp/test/CMakeLists.txt#L53 And https://github.com/rapidsai/cuml/blob/6f42ced4d6f84cc04ea27f8940449d57fa08ab51/cpp/test/CMakeLists.txt#L19

Reply:
Ah, yes that is much cleaner indeed. I've adjusted the cmake accordingly.